sympy/polys/orthopolys.py

```diff
@@ -72,11 +72,74 @@ def dup_chebyshevt(n, K):
     """Low-level implementation of Chebyshev polynomials of the first kind."""
     if n < 1:
         return [K.one]
+    # When n is small, it is faster to directly calculate the recurrence relation.
+    if n < 64: # The threshold serves as a heuristic
+        return _dup_chebyshevt_rec(n, K)
+    return _dup_chebyshevt_prod(n, K)
+
+def _dup_chebyshevt_rec(n, K):
+    r""" Low-level implementation of Chebyshev polynomials of the first kind using a recurrence relation.
+
+    Explanation
+    ===========
+
+    Chebyshev polynomials of the first kind are defined by the recurrence relation:
+
+    .. math::
+        T_0(x) &= 1\\
+        T_1(x) &= x\\
+        T_n(x) &= 2xT_{n-1}(x) - T_{n-2}(x)
+
+    This function calculates the Chebyshev polynomial of the first kind using the above recurrence relation.
+
+    Parameters
+    ==========
+
+    n : int
+        n is a nonnegative integer.
+    K : domain
+
+    """
     m2, m1 = [K.one], [K.one, K.zero]
-    for i in range(2, n+1):
+    for _ in range(n - 1):
         m2, m1 = m1, dup_sub(dup_mul_ground(dup_lshift(m1, 1, K), K(2), K), m2, K)
     return m1
 
+def _dup_chebyshevt_prod(n, K):
+    r""" Low-level implementation of Chebyshev polynomials of the first kind using a product expression.
+
+    Explanation
+    ===========
+
+    Chebyshev polynomials of the first kind are expressed by the following product formulas:
+
+    .. math::
+        T_{2n}(x) &= 2T_n^2(x) - 1\\
+        T_{2n+1}(x) &= 2T_{n+1}(x)T_n(x) - x
+
+    Consequently, to calculate `(T_n(x), T_{n+1}(x))`, we need to know
+    either `(T_{n/2}(x), T_{n/2+1}(x))` or `(T_{(n-1)/2}(x), T_{(n-1)/2+1}(x))`.
+    Starting with `(T_1(x), T_2(x))`, we can derive `(T_n(x), T_{n+1}(x))` for any integer n.
+    This algorithm, which requires only `O(\log n)` computations,
+    is more efficient than using the standard recurrence relation for large n.
+
+    Parameters
+    ==========
+
+    n : int
+        n is a nonnegative integer.
+    K : domain
+
+    """
```

Comment:
I don't think that these functions need such long docstrings especially if internal only (does this even appear in the html docs?).

I would shorten this to maybe
```python
def _dup_chebyshevt_prod(n, K):
    r"""Chebyshev T polynomials using recursive products

    Explanation
    ===========

    Computes Chebyshev T polynomials using

    .. math::
        T_{2n}(x) &= 2T_n^2(x) - 1\\
        T_{2n+1}(x) &= 2T_{n+1}(x)T_n(x) - x

    This is faster than ``_dup_chebyshevt_rec`` for large ``n``.

    Parameters
    ==========
    n : int (nonegative)
    K : domain
    """
```

Reply:
I think it would be appreciated to have as much documentation as possible, even for functions for internal use.

Reply:
I agree that documentation is good but "as much as possible" is not right: there has to be some limit. At some point it is better to understand what is happening by looking at the code than the documentation but some documentation of internals also helps with that. Right now functions in the polys module have too little documentation but it is definitely also possible to go too far in the other direction especially here when we have multiple functions that all take the same arguments and return the exact same thing.

Also we need to consider who this documentation is for i.e. who is going to read about the internal functions in this module and we can we reasonably expect them to know? Here I would say that the reader is either a SymPy developer or someone else who understands the internals enough to be able to use such things.

Here showing the recursion relation is good because someone can see that that is the basis of the correctness of the code. It is also the significant thing that is different about this function vs the other two functions that also return the same thing. Explaining that it is faster than the other function in some cases is good because that gives the background for the purpose of the code and again contrasts it with the two other functions.

Otherwise I don't think that (in this particular case) more is needed. If someone wants to measure/analyse the performance/complexity then they can but we should only mention here the main things needed to understood purpose/correctness or something that is very not obvious.

In this particular case also the comment about `O(log(n))` computations is misleading since it implicitly compares one method that uses only addition and scalar multiplication with another that uses polynomial multiplication (a more expensive operation).